go/vt/vtgate/tabletgateway.go

```diff
@@ -231,10 +258,22 @@ func (gw *TabletGateway) withRetry(ctx context.Context, target *querypb.Target,
 				defer retryDone()
 				bufferedOnce = true
 			}
+
+			if bufferErr != nil {
```

Comment:
Could you provide some context on this change in the order of the logic here? We are checking retryDone, before checking the error. 

I'm not familiarized with the details of the logic here, but this chance caught my attention.

Reply:
This is a small corner case fix: there are now cases when the buffering code returns both an error and a cancelation function. It's important to `defer` the cancelation function whenever it's returned, even if we also have an error and must exit the function right away -- not defering the function would cause a (tiny) memory leak.

Reply:
Ah cool. Makes sense as well.